File: cpp-equivalents/surface/surface.rs
use std::ops::Drop;

#[derive(Clone, Copy)]
struct Dummy;
impl Dummy {
	pub unsafe fn destroy_surface(&self, _other: Dummy) {
		println!("Destroying dummy");
	}
}
type Loader = Dummy;
type SurfaceKHR = Dummy;

#[derive(Debug)]
struct Window {
	a: i32
}
impl Window {
	pub fn new(a: i32) -> Self {
		Window { a }
	}

	// There is no move constructor, all types are moveable
}
impl Drop for Window {
	fn drop(&mut self) {
		println!("Destroying window {}", self.a);
	}
}

struct Surface<W: std::fmt::Debug> {
	window: Option<W>,
	loader: Loader,
	surface: SurfaceKHR
}
impl<W: std::fmt::Debug> Surface<W> {
	pub fn new(window: W) -> Self {
		Surface {
			window: Some(window),
			loader: Dummy,
			surface: Dummy
		}
	}

	pub fn destroy_without_window(mut self) -> W {
		self.window.take().unwrap()
	}
}
impl<W: std::fmt::Debug> Drop for Surface<W> {
	fn drop(&mut self) {
		println!("Destroying surface with window {:?}", self.window);

		unsafe {
			self.loader.destroy_surface(self.surface);
		}
	}
}

fn main() {
	println!(">> Moving out");
	{
		let original = Window::new(1);
		println!("original {:?}\n", original);

		let surface = Surface::new(original);
		// println!("original {:?}", original); // Compiler error, original was moved
		println!("in surface {:?}\n", surface.window);

		let moved = surface.destroy_without_window();
		// println!("original {:?}", original); // Compiler error, original was moved
		// println!("in surface {:?}", surface.window); // Compiler error, surface was moved
		println!("moved out {:?}\n", moved);
	}

	println!("\n>>Not moving out");
	{
		let original = Window::new(2);
		println!("original {:?}\n", original);

		let surface = Surface::new(original);
		println!("in surface {:?}\n", surface.window);
	}
}

File: codes/cpp-equivalents/surface/surface.cpp
#include <algorithm>
#include <iostream>

class Dummy {
	public:
		void destroy_surface(Dummy other) {
			std::cout << "Destroying dummy" << std::endl; 
		}
};
typedef Dummy Loader;
typedef Dummy SurfaceKHR;

class Window {
	public:
		Window(int a) : a(a) {}
		Window(Window&& move) {
			std::cout << "Moving window " << move.a << std::endl;
			this->a = move.a;
			move.a = -move.a;
		}

		~Window() {
			std::cout << "Destroying window " << this->a << std::endl;
		}

	// private:
		int a;
};

template<class W>
class Surface {
	public:
		Surface(W window) : window(std::move(window)) {}
		
		~Surface() {
			std::cout << "Destroying surface with window " << this->window.a << std::endl;

			this->loader.destroy_surface(this->surface);
		}

		W destroy_without_window() {
			return std::move(this->window);
		}

	// private:
		W window;

		Loader loader;
		SurfaceKHR surface;
};

int main() {
	std::cout << ">> Moving out" << std::endl;
	{
		Window original(1);
		std::cout << "original " << original.a << std::endl << std::endl;

		Surface<Window> surface(std::move(original));
		std::cout << "original " << original.a << std::endl;
		std::cout << "in surface " << surface.window.a << std::endl << std::endl;

		Window moved = surface.destroy_without_window();

		std::cout << "original " << original.a << std::endl;
		std::cout << "in surface " << surface.window.a << std::endl;
		std::cout << "moved out " << moved.a << std::endl << std::endl;
	}

	std::cout << std::endl << ">> Not moving out" << std::endl;
	{
		Window original(2);
		std::cout << "original " << original.a << std::endl << std::endl;

		Surface<Window> surface(std::move(original));
		std::cout << "original " << original.a << std::endl;
		std::cout << "in surface " << surface.window.a << std::endl << std::endl;
	}
}